File: 10_ca_provincia_organismo.py
"""
10 — Correspondence analysis: territorial supply x state demand
===============================================================
Project: Who Sells to the State (2026_2). GAN V3 remedy — the central
relational instrument: provinces (supply positions) and buyer types (demand
poles) in one geometric space, per era; plus rubro x buyer-type.

Input:  data/processed/adjudicaciones_tipo.parquet
Output: tables/tab_ca_coords_provincia_tipo.csv
        figures/fig_ca_provincia_tipo.png
        figures/fig_ca_rubro_tipo.png

Usage:  python 10_ca_provincia_organismo.py
"""
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import prince

PROJECT = Path(__file__).parent
PROC = PROJECT / "data" / "processed"
TAB = PROJECT / "tables"
FIG = PROJECT / "figures"

NEA = {"Misiones", "Corrientes", "Chaco", "Formosa"}
ERAS = ["macri", "fernandez", "milei"]

adj = pd.read_parquet(PROC / "adjudicaciones_tipo.parquet")
adj = adj[adj["es_nuevo"] & adj["provincia"].notna()]

# ── CA provincia x tipo_organismo, pooled and per era ────────────────────────
def run_ca(table):
    ca = prince.CA(n_components=2, random_state=42).fit(table)
    return (ca, ca.row_coordinates(table), ca.column_coordinates(table))


coords_all = []
fig, axes = plt.subplots(1, 3, figsize=(16, 5.2))
for ax, era in zip(axes, ERAS):
    xt = pd.crosstab(adj.loc[adj["era"] == era, "provincia"],
                     adj.loc[adj["era"] == era, "tipo_organismo"])
    xt = xt[xt.sum(axis=1) >= 30]  # provinces with enough mass
    ca, rows, cols = run_ca(xt)
    inertia = ca.percentage_of_variance_
    for prov, r in rows.iterrows():
        is_nea = prov in NEA
        ax.scatter(r[0], r[1], s=60 if is_nea else 25,
                   color="#d62728" if is_nea else "#7fb3d5", zorder=3)
        ax.annotate(prov, (r[0], r[1]), fontsize=6.5,
                    color="#a00" if is_nea else "#333",
                    xytext=(3, 3), textcoords="offset points")
        coords_all.append({"era": era, "punto": prov, "clase": "provincia",
                           "dim1": r[0], "dim2": r[1]})
    for tipo, r in cols.iterrows():
        ax.scatter(r[0], r[1], marker="^", s=120, color="#2ca02c", zorder=4)
        ax.annotate(tipo, (r[0], r[1]), fontsize=8, weight="bold",
                    color="#155d15", xytext=(4, -8), textcoords="offset points")
        coords_all.append({"era": era, "punto": tipo, "clase": "tipo_organismo",
                           "dim1": r[0], "dim2": r[1]})
    ax.axhline(0, color="grey", lw=0.5)
    ax.axvline(0, color="grey", lw=0.5)
    ax.set_title(f"{era} (dim1 {inertia[0]:.0f}%, dim2 {inertia[1]:.0f}%)")
fig.suptitle("CA: provincias (oferta territorial) × tipos de organismo (demanda estatal)")
fig.tight_layout()
fig.savefig(FIG / "fig_ca_provincia_tipo.png", dpi=200)
plt.close(fig)
pd.DataFrame(coords_all).to_csv(TAB / "tab_ca_coords_provincia_tipo.csv", index=False)

# ── CA rubro x tipo_organismo (pooled) ───────────────────────────────────────
top_rubros = adj["rubro_principal"].value_counts().head(20).index
xt2 = pd.crosstab(adj.loc[adj["rubro_principal"].isin(top_rubros), "rubro_principal"],
                  adj.loc[adj["rubro_principal"].isin(top_rubros), "tipo_organismo"])
ca2, rows2, cols2 = run_ca(xt2)
fig, ax = plt.subplots(figsize=(9, 7))
for rub, r in rows2.iterrows():
    ax.scatter(r[0], r[1], s=20, color="#7f7f7f")
    ax.annotate(str(rub)[:22], (r[0], r[1]), fontsize=6.5, color="#444",
                xytext=(3, 3), textcoords="offset points")
for tipo, r in cols2.iterrows():
    ax.scatter(r[0], r[1], marker="^", s=140, color="#2ca02c", zorder=4)
    ax.annotate(tipo, (r[0], r[1]), fontsize=9, weight="bold", color="#155d15")
ax.axhline(0, color="grey", lw=0.5)
ax.axvline(0, color="grey", lw=0.5)
ax.set_title(f"CA: rubros × tipos de organismo "
             f"(dim1 {ca2.percentage_of_variance_[0]:.0f}%, "
             f"dim2 {ca2.percentage_of_variance_[1]:.0f}%)")
fig.tight_layout()
fig.savefig(FIG / "fig_ca_rubro_tipo.png", dpi=200)
plt.close(fig)

# print NEA positions per era vs the seguridad pole
print("Posición NEA vs polo seguridad_defensa (dim1, por era):")
cdf = pd.DataFrame(coords_all)
for era in ERAS:
    d = cdf[cdf["era"] == era].set_index("punto")
    if "seguridad_defensa" not in d.index:
        continue
    pole = d.loc["seguridad_defensa", "dim1"]
    neap = d[d.index.isin(NEA)]["dim1"]
    print(f"  {era:10s} polo_seg={pole:+.2f} | " +
          " ".join(f"{p}={v:+.2f}" for p, v in neap.items()))
print("OK -> tables/, figures/")

File: 10_ca_provincia_organismo.R
# 10 — Correspondence analysis: territorial supply x state demand
# (canonical R port of 10_ca_provincia_organismo.py; FactoMineR::CA).
# Axis signs are arbitrary; comparisons with the Python pass are structural.
# Run: Rscript 10_ca_provincia_organismo.R

source("theme_house.R", chdir = TRUE)
suppressPackageStartupMessages({library(FactoMineR); library(ggrepel)})

# window: before 2019 the platform holds a different set of buyers, so an era
# comparison across the rollout would compare states, not markets. The Macri
# era is therefore represented by 2019 alone (see 20_platform_stability.R).
adj <- read_parquet(file.path(DIR_PROC, "adjudicaciones_tipo.parquet")) |>
  filter(es_nuevo, !is.na(provincia), between(ejercicio, WIN0, WIN1))

run_ca <- function(tab) CA(tab, ncp = 2, graph = FALSE)

coords_all <- list()
plots <- list()
for (e in ERAS) {
  xt <- adj |> filter(era == e) |> count(provincia, tipo_organismo) |>
    pivot_wider(names_from = tipo_organismo, values_from = n,
                values_fill = 0) |>
    as.data.frame()
  rownames(xt) <- xt$provincia; xt$provincia <- NULL
  xt <- xt[rowSums(xt) >= 30, ]
  ca <- run_ca(as.matrix(xt))
  rows <- as.data.frame(ca$row$coord)[, 1:2]
  cols <- as.data.frame(ca$col$coord)[, 1:2]
  names(rows) <- names(cols) <- c("dim1", "dim2")
  inertia <- ca$eig[1:2, 2]
  coords_all[[e]] <- bind_rows(
    rows |> mutate(era = e, punto = rownames(rows), clase = "provincia"),
    cols |> mutate(era = e, punto = rownames(cols), clase = "tipo_organismo")) |>
    mutate(inercia1 = round(inertia[1], 1), inercia2 = round(inertia[2], 1))
  d <- coords_all[[e]] |>
    mutate(nea = clase == "provincia" & punto %in% NEA)
  plots[[e]] <- ggplot(d, aes(dim1, dim2)) +
    geom_hline(yintercept = 0, colour = "grey70", linewidth = 0.3) +
    geom_vline(xintercept = 0, colour = "grey70", linewidth = 0.3) +
    geom_point(data = subset(d, clase == "provincia"),
               aes(colour = nea, size = nea)) +
    geom_point(data = subset(d, clase == "tipo_organismo"),
               shape = 17, size = 3, colour = "#2ca02c") +
    geom_text_repel(aes(label = punto,
                        fontface = ifelse(clase == "tipo_organismo",
                                          "bold", "plain")),
                    size = 2.3, max.overlaps = 30, seed = 42) +
    scale_colour_manual(values = c(`TRUE` = "#d62728", `FALSE` = "#7fb3d5"),
                        guide = "none") +
    scale_size_manual(values = c(`TRUE` = 2.6, `FALSE` = 1.6), guide = "none") +
    labs(title = sprintf("%s (dim1 %.0f%%, dim2 %.0f%%)", e,
                         inertia[1], inertia[2]),
         x = NULL, y = NULL) +
    theme_house(9)
}
tab <- bind_rows(coords_all) |>
  select(era, punto, clase, dim1, dim2, inercia1, inercia2)
write_tab(tab, "tab_ca_coords_provincia_tipo.csv")

library(patchwork)
p <- plots[[1]] + plots[[2]] + plots[[3]] +
  patchwork::plot_annotation(
    title = "CA: provincias (oferta territorial) × tipos de organismo (demanda estatal)")
save_fig(p, "../_archive/figures_working/fig_ca_provincia_tipo.png",
         width = 16, height = 5.4)

# ── CA rubro x tipo (pooled) ─────────────────────────────────────────────────
top_rubros <- adj |> count(rubro_principal, sort = TRUE) |>
  slice_head(n = 20) |> pull(rubro_principal)
xt2 <- adj |> filter(rubro_principal %in% top_rubros) |>
  count(rubro_principal, tipo_organismo) |>
  pivot_wider(names_from = tipo_organismo, values_from = n, values_fill = 0) |>
  as.data.frame()
rownames(xt2) <- xt2$rubro_principal; xt2$rubro_principal <- NULL
ca2 <- run_ca(as.matrix(xt2))
r2 <- as.data.frame(ca2$row$coord)[, 1:2] |>
  setNames(c("dim1", "dim2")) |>
  mutate(punto = substr(rownames(ca2$row$coord), 1, 22), clase = "rubro")
c2 <- as.data.frame(ca2$col$coord)[, 1:2] |>
  setNames(c("dim1", "dim2")) |>
  mutate(punto = rownames(ca2$col$coord), clase = "tipo")
p2 <- ggplot(bind_rows(r2, c2), aes(dim1, dim2)) +
  geom_hline(yintercept = 0, colour = "grey70", linewidth = 0.3) +
  geom_vline(xintercept = 0, colour = "grey70", linewidth = 0.3) +
  geom_point(aes(shape = clase, colour = clase, size = clase)) +
  geom_text_repel(aes(label = punto, fontface = ifelse(clase == "tipo",
                                                       "bold", "plain")),
                  size = 2.3, max.overlaps = 40, seed = 42) +
  scale_shape_manual(values = c(rubro = 16, tipo = 17), guide = "none") +
  scale_colour_manual(values = c(rubro = "#7f7f7f", tipo = "#2ca02c"),
                      guide = "none") +
  scale_size_manual(values = c(rubro = 1.6, tipo = 3), guide = "none") +
  labs(title = sprintf("CA: rubros × tipos de organismo (dim1 %.0f%%, dim2 %.0f%%)",
                       ca2$eig[1, 2], ca2$eig[2, 2]),
       x = NULL, y = NULL) +
  theme_house()
save_fig(p2, "../_archive/figures_working/fig_ca_rubro_tipo.png",
         width = 9, height = 7)

cat("\nDistancia NEA al polo seguridad_defensa (por era):\n")
for (e in ERAS) {
  d <- coords_all[[e]]
  pole <- d |> filter(punto == "seguridad_defensa")
  neap <- d |> filter(clase == "provincia", punto %in% NEA)
  dist <- sqrt((neap$dim1 - pole$dim1)^2 + (neap$dim2 - pole$dim2)^2)
  cat(sprintf("  %-10s %s\n", e,
              paste(sprintf("%s=%.2f", neap$punto, dist), collapse = "  ")))
}
